Add poetry package management (#660)
eriklindernoren/PyTorch-YOLOv3: README.md

A minimal PyTorch implementation of YOLOv3, with support for training, inference and evaluation.

## Installation
### Installing from source

For normal training and evaluation we recommend installing the package from source using a poetry virtual enviroment.

```bash
git clone https://github.com/eriklindernoren/PyTorch-YOLOv3
cd PyTorch-YOLOv3/
pip3 install poetry --user
poetry install
```

> You need join the virtual enviroment by runing `poetry shell` in this directory before running any of the following commands without the `poetry run` prefix. Also have a look at the other installing method if you want to use the commands everywhere without opening a poetry shell.

Reviewer: ```suggestion
You need to join the virtual enviroment by runing `poetry shell` in this directory before running any of the following commands without the `poetry run` prefix.
Also have a look at the other installing method, if you want to use the commands everywhere without opening a poetry-shell.
```
I think, it is a little bit unclear, what the difference between `shell` and `run` is, and why to use `shell`, when not using `run`.

#### Download pretrained weights

```bash

> cd weights/
> bash download_weights.sh

Reviewer: ```suggestion
./weights/download_weights.sh
```
The bash command is not needed, as the file has a shebang, but I don't know, if this file has the executable flag...

```

#### Download COCO

```bash

> cd data/
> bash get_coco_dataset.sh

Reviewer: ```suggestion
./data/get_coco_dataset.sh
```
Same as above

```

### Install via pip

> This installation method is less isolated and may conflict with other packages. It also only includes the code. Weights and the COCO dataset need to be installed as stated above. With this beeing said it is a good method if you want to use this package as dependency in another python project. See __API__ for further information regarding the packages api. It also enables the cli tools `yolo-detect`, `yolo-train`, and `yolo-test` everywhere without any additional commands.

Reviewer: ```suggestion
This installation method is recommended, if you want to use this package as dependency in another python project.
This method only includes the code, is less isolated and may conflict with other packages.
Weights and the COCO dataset need to be downloaded as stated above.
See __API__ for further information regarding the packages API.
It also enables the CLI tools `yolo-detect`, `yolo-train`, and `yolo-test` everywhere without any additional commands.
```
A few things are unclear to me:

- Why would this conflict with other packages? Is there anything special, compared to other published packages, that makes this more risky?
- Weights and dataset can not be `installed`or downloaded as stated above. The package is installed in a other directory, which also should not contain data.
- What do you mean by `enabling`cli tools?

Author: * Not really, but if the installs another project without a venv with numpy==4.2.0 and we require numpy==6.9.0 it may conflicts and results in undefined behavior.
* The code is installed in a "hidden" system folder so if the user runs e.g. `yolo-train` he needs to use the correct path for the weights file. This shouldn't be a problem if he runs the command in the root of the cloned git repo because that would match the default paths.
* You are able to use them just like `git` or any other command in the `PATH`without sourcing your venv aka. run `poetry shell`or something simmilar.

```bash
pip3 install pytorchyolo --user
```

## Test

> Evaluates the model on COCO test dataset.

Reviewer: ```suggestion
Evaluates the model on COCO test dataset.
To download this dataset as well as weights, see above.
```

poetry install
```

You need to join the virtual enviroment by runing `poetry shell` in this directory before running any of the following commands without the `poetry run` prefix.
Also have a look at the other installing method, if you want to use the commands everywhere without opening a poetry-shell.

#### Download pretrained weights

```bash
./weights/download_weights.sh
```

#### Download COCO

```bash
./data/get_coco_dataset.sh
```

### Install via pip

> This installation method is recommended, if you want to use this package as dependency in another python project.

Reviewer: ```suggestion
This installation method is recommended, if you want to use this package as a dependency in another python project.
```

* Run the command below
* Go to http://localhost:6006/

```bash
poetry run tensorboard --logdir='logs' --port=6006
```

Storing the logs on a slow drive possibly leads to a significant training speed decrease.

You can adjust the log directory using `--logdir <path>` when running `tensorboard` and `yolo-train`.

## Train on Custom Dataset

#### Custom model
Run the commands below to create a custom model definition, replacing `<num-classes>` with the number of classes in your dataset.

```bash

> cd config/                                # Navigate to config dir
> bash create_custom_model.sh <num-classes> # Will create custom model 'yolov3-custom.cfg'

Reviewer: ```suggestion
./config/create_custom_model.sh <num-classes>  # Will create custom model 'yolov3-custom.cfg'
```
See above

Add `--pretrained_weights weights/darknet53.conv.74` to train using a backend pretrained on ImageNet.

## API

You are able to import the modules of this repo in your own project if you install the pip package `pytorchyolo`.

An example prediction call from a simple OpenCV python script would look like this:

```python
import cv2
from pytorchyolo import detect, models

# Load the YOLO model
model = models.load_model(
  "<PATH_TO_YOUR_CONFIG_FOLDER>/yolov3.cfg", 
  "<PATH_TO_YOUR_WEIGHTS_FOLDER>/yolov3.weights")

# Load the image as an numpy array
img = cv2.imread("<PATH_TO_YOUR_IMAGE>")

# Runs the YOLO model on the image 
boxes = detect.detect_image(model, img)

print(boxes)
```

> For more advanced usage look at the method doc strings.

Reviewer: ```suggestion
For more advanced usage look at the method's doc strings.
```
